drivers/hmc5883l/hmc5883l.c

```diff
@@ -0,0 +1,338 @@
+/*
+ * Copyright (C) 2018 Gunar Schorcht
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_hmc5883l
+ * @brief       Device driver for the Honeywell HMC5883L 3-axis digital compass
+ * @author      Gunar Schorcht <gunar@schorcht.net>
+ * @file
+ * @{
+ */
+
+#include <string.h>
+#include <stdlib.h>
+
+#include "hmc5883l_regs.h"
+#include "hmc5883l.h"
+
+#include "log.h"
+#include "xtimer.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#if ENABLE_DEBUG
+
+#define ASSERT_PARAM(cond) \
+    if (!(cond)) { \
+        DEBUG("[hmc5883l] %s: %s\n", \
+              __func__, "parameter condition (" # cond ") not fulfilled"); \
+        assert(cond); \
+    }
+
+#define DEBUG_DEV(f, d, ...) \
+        DEBUG("[hmc5883l] %s i2c dev=%d addr=%02x: " f "\n", \
+              __func__, d->params.dev, d->params.addr, ## __VA_ARGS__);
+
+#else /* ENABLE_DEBUG */
+
+#define ASSERT_PARAM(cond) assert(cond);
+#define DEBUG_DEV(f, d, ...)
+
+#endif /* ENABLE_DEBUG */
+
+#define ERROR_DEV(f, d, ...) \
+        LOG_ERROR("[hmc5883l] %s i2c dev=%d addr=%02x: " f "\n", \
+                  __func__, d->params.dev, d->params.addr, ## __VA_ARGS__);
+
+#define EXEC_RET(f, r) \
+    if ((r = f) != HMC5883L_OK) { \
+        DEBUG("[hmc5883l] %s: error code %d\n", __func__, res); \
+        return res; \
+    }
+
+#define EXEC_RET_CODE(f, r, c) \
+    if ((r = f) != HMC5883L_OK) { \
+        DEBUG("[hmc5883l] %s: error code %d\n", __func__, res); \
+        return c; \
+    }
+
+/** Forward declaration of functions for internal use */
+
+static int _is_available(const hmc5883l_t *dev);
+
+static uint8_t _get_reg_bit(uint8_t byte, uint8_t mask);
+static void _set_reg_bit(uint8_t *byte, uint8_t mask, uint8_t bit);
+static int _reg_read(const hmc5883l_t *dev, uint8_t reg, uint8_t *data, uint16_t len);
+static int _reg_write(const hmc5883l_t *dev, uint8_t reg, uint8_t *data, uint16_t len);
+
+int hmc5883l_init(hmc5883l_t *dev, const hmc5883l_params_t *params)
+{
+    int res = HMC5883L_OK;
+
+    ASSERT_PARAM(dev != NULL);
+    ASSERT_PARAM(params != NULL);
+    DEBUG_DEV("params=%p", dev, params);
+
+    /* init sensor data structure */
+    dev->params = *params;
+
+    /* check availability of the sensor */
+    EXEC_RET(_is_available(dev), res)
+
+    uint8_t cfg_a = 0;
+    uint8_t cfg_b = 0;
+    uint8_t mode = 0;
+
+    /* set configuration register A and B */
+    _set_reg_bit(&cfg_a, HMC5883L_REG_CFG_A_MA, params->meas_avg);
+    _set_reg_bit(&cfg_a, HMC5883L_REG_CFG_A_MS, params->meas_mode);
+    _set_reg_bit(&cfg_a, HMC5883L_REG_CFG_A_DO, params->dor);
+    _set_reg_bit(&cfg_b, HMC5883L_REG_CFG_B_GN, params->gain);
```

Comment:
This seems much too complicated to me. I suggest to do something like
```c
uint8_t tmp = (params->meas->avg | params->meas_mode | params->dor);
_reg_write(dev->dev, dev->addr, HMC5883L_REG_CFG_A, tmp, 0); // of course test return value for error...
tmp = params->gain;
_reg_write(..., HMC5883L_REG_CFG_B, tmp, 0);
[...]
```
Now the only thing that is needed to be changed is to assing more specifig values to the config param enums, e.g.:
```c
typedef enum {
    HMC5883L_DOR_0_75 = 0x00,  /**< 0.75 Hz           */
    HMC5883L_DOR_1_5 = 0x04,       /**<  1.5 Hz           */
    HMC5883L_DOR_3 = 0x08,         /**<    3 Hz           */
    HMC5883L_DOR_7_5 = 0xc,       /**<  7.5 Hz           */
    HMC5883L_DOR_15 = 0x10,        /**<   15 Hz (default) */
    HMC5883L_DOR_30 = 0x14,        /**<   30 Hz           */
    HMC5883L_DOR_75 = 0x18,        /**<   75 Hz           */
} hmc5883l_dor_t;
```
This should be much easier to read and I would think it should also be more efficient?!

Reply:
ORing fix values is definitely more efficient than to shift them bit-wise in a loop. Since they are only set during setup or when powering off/on, performance shouldn't be a problem.

But from my point of view, using fixed values instead of masks and shifts is more error-prone in programming. I had just to learn it when I tried to change these values it in this driver